core/key.go

```diff
@@ -318,6 +351,101 @@ func (k DataStoreKey) ToPrimaryDataStoreKey() PrimaryDataStoreKey {
 	}
 }
 
+// NewIndexDataStoreKey creates a new IndexDataStoreKey from a string as best as it can,
+// splitting the input using '/' as a field deliminator.  It assumes
+// that the input string is in the following format:
+//
+// /[CollectionID]/[IndexID]/[FieldID](/[FieldID]...)
+//
+// Any properties before the above (assuming a '/' deliminator) are ignored
+func NewIndexDataStoreKey(key string) (IndexDataStoreKey, error) {
+	indexKey := IndexDataStoreKey{}
+	if key == "" {
+		return indexKey, ErrEmptyKey
```

Comment:
nitpick: I don't like returning `indexKey` here and on similar error lines below, its technical-correctness is somewhat dependent on other lines (including future modifications) not mutating the `indexKey` var, which is actually done a bunch of times later in the func.  

Technical-correctness aside, I see it as conceptually incorrect, as we dont actually want to return `indexKey`, it is just that it happens to currently have the same (empty) value that we do want to return.

Doing this will also allow you to move the `indexKey` declaration lower into the function, closer to where it is actually used, and reducing the scope of its existence.